stack/__main__.py

```diff
@@ -0,0 +1,391 @@
+    """Returns the bucket name for the given dataset."""
+    return f'cpg-{dataset}-{kind}'
+
+
+def create_bucket(name: str, **kwargs) -> gcp.storage.Bucket:
+    """Returns a new GCS bucket."""
+    return gcp.storage.Bucket(
+        name,
+        name=name,
+        location=REGION,
+        versioning=gcp.storage.BucketVersioningArgs(enabled=True),
+        labels={'bucket': name},
+        **kwargs,
+    )
+
+
+undelete_rule = gcp.storage.BucketLifecycleRuleArgs(
+    action=gcp.storage.BucketLifecycleRuleActionArgs(type='Delete'),
+    condition=gcp.storage.BucketLifecycleRuleConditionArgs(
+        age=30, with_state='ARCHIVED'
+    ),
+)
+
+upload_bucket = create_bucket(bucket_name('upload'), lifecycle_rules=[undelete_rule])
+
+# The Cloud Identity API is required for creating access groups.
+cloudidentity = gcp.projects.Service(
+    'cloudidentity-service',
+    service='cloudidentity.googleapis.com',
+    disable_on_destroy=False,
+)
+
+upload_account = gcp.serviceaccount.Account(
+    'upload-service-account',
+    account_id='upload',
+    display_name='upload',
+    opts=pulumi.resource.ResourceOptions(depends_on=[cloudidentity]),
+)
+
+gcp.storage.BucketIAMMember(
+    'upload-permissions-viewer',
+    bucket=upload_bucket.name,
+    role='roles/storage.objectViewer',
+    member=pulumi.Output.concat('serviceAccount:', upload_account.email),
+)
+
+gcp.storage.BucketIAMMember(
+    'upload-permissions-creator',
+    bucket=upload_bucket.name,
+    role='roles/storage.objectCreator',
+    member=pulumi.Output.concat('serviceAccount:', upload_account.email),
+)
+
+archive_bucket = create_bucket(
+    bucket_name('archive'),
+    lifecycle_rules=[
+        gcp.storage.BucketLifecycleRuleArgs(
+            action=gcp.storage.BucketLifecycleRuleActionArgs(
+                type='SetStorageClass', storage_class='ARCHIVE'
+            ),
+            condition=gcp.storage.BucketLifecycleRuleConditionArgs(age=archive_age),
+        ),
+        undelete_rule,
+    ],
+)
+
+main_bucket = create_bucket(bucket_name('main'), lifecycle_rules=[undelete_rule])
+analysis_bucket = create_bucket(
+    bucket_name('analysis'), lifecycle_rules=[undelete_rule]
+)
+test_bucket = create_bucket(bucket_name('test'), lifecycle_rules=[undelete_rule])
+
+temporary_bucket = create_bucket(
+    bucket_name('temporary'),
+    lifecycle_rules=[
+        gcp.storage.BucketLifecycleRuleArgs(
+            action=gcp.storage.BucketLifecycleRuleActionArgs(type='Delete'),
+            condition=gcp.storage.BucketLifecycleRuleConditionArgs(
+                age=30, with_state='LIVE'
+            ),
+        ),
+        undelete_rule,
+    ],
+)
+
+
+def group_mail(kind: str) -> str:
+    """Returns the email address of a permissions group."""
+    return f'{dataset}-{kind}@{DOMAIN}'
+
+
+def create_group(mail: str) -> gcp.cloudidentity.Group:
+    """Returns a new Cloud Identity group for the given email address."""
+    name = mail.split('@')[0]
```

Comment:
Should we process the name further, by maybe making it lowercase with `.lower()`?

Reply:
The name here is just a Pulumi identifier that's displayed in the plan. So there's no need to do a lot of processing, it just was a bit long when having the domain attached to it.